XEP-0313 MAM support.

thibaultmarin opened the pull request:
This is an attempt at implementing support for XEP-0313 Message Archive Management (http://xmpp.org/extensions/xep-0313.html).

I am testing this with a prosody server (Prosody trunk nightly build 732 (2016-12-19, 8efd51667622)) and mercurial prosody-modules (https://hg.prosody.im/prosody-modules/ changeset 2462:81127dcdb326).

It would be useful if people could test this and report.

thibaultmarin committed:
XEP-0313 MAM support.

DamienCassou commented on jabber-history.el:
```diff
@@ -49,6 +50,12 @@ Default is nil, cause MUC logging may be i/o-intensive."
   :type 'boolean
   :group 'jabber-history)
 
+(defcustom jabber-history-mam nil
```
shouldn't that be a per-account option instead?

DamienCassou commented on jabber-mam.el:
```diff
@@ -0,0 +1,206 @@
+  (nreverse jabber-mam-results))
+
+(provide 'jabber-mam)
+;;; jabber-mam.el ends here
```
Functions in these files are quite long. Your code would gain readability if some smaller functions were extracted from the large ones.

DamienCassou reviewed (commented):
Good job. I hope a maintainer will look at it.

thibaultmarin committed:
Split code into multiple functions.

thibaultmarin committed:
Fix bug in end of response processing.
Prevent hang on error.

thibaultmarin commented:
Hi, thanks for the review.

> DamienCassou commented on this pull request.
> +(defcustom jabber-history-mam nil
>
> shouldn't that be a per-account option instead?

The current scope follows that of the `jabber-history-enabled' option.
Would it make sense to keep both options with the same scope in this
pull request, and possibly make the options account-specific at a later
time?  I'd rather isolate the MAM implementation from other
improvements, if that makes sense.

That being said, the code would hang when MAM is enabled but not
supported by a server, so I have implemented a tentative fix for that.


> Functions in these files are quite long. Your code would gain readability if some smaller functions were extracted from the large ones.

I updated the code, splitting functions into smaller components.  I hope
it improves readability.

thibaultmarin committed:
Decouple jabber-history-enabled and jabber-history-mam (get history from mam when both set)

Commit:
Update to mam:2, whitespace fixes

Neustradamus commented:
@thibaultmarin: Can you update your PR with latest XEP-0313 version?
- https://xmpp.org/extensions/xep-0313.html

Currently it is 1.1.3 (2025-04-09).

Thanks in advance.
